src/renderers/WebGLRenderer.js

```diff
@@ -2055,64 +2045,74 @@ THREE.WebGLRenderer = function ( parameters ) {
 		} else if ( material.program.code !== code ) {
 
 		    // changed glsl or parameters
-		    deallocateMaterial( material );
+			deallocateMaterial( material );
+			material.program = undefined;
 
-		} else {
+		}
 
-            // same glsl and parameters
-		    return;
+		var program;
 
-		}
+		if ( !material.program ) {
 
-		if ( shaderID ) {
+			if ( shaderID ) {
 
-			var shader = THREE.ShaderLib[ shaderID ];
+				var shader = THREE.ShaderLib[shaderID];
 
-			material.__webglShader = {
-				uniforms: THREE.UniformsUtils.clone( shader.uniforms ),
-				vertexShader: shader.vertexShader,
-				fragmentShader: shader.fragmentShader
-			}
+				material.__webglShader = {
+					uniforms: THREE.UniformsUtils.clone( shader.uniforms ),
+					vertexShader: shader.vertexShader,
+					fragmentShader: shader.fragmentShader
+				}
 
-		} else {
+			} else {
 
-			material.__webglShader = {
-				uniforms: material.uniforms,
-				vertexShader: material.vertexShader,
-				fragmentShader: material.fragmentShader
-			}
+				material.__webglShader = {
+					uniforms: material.uniforms,
+					vertexShader: material.vertexShader,
+					fragmentShader: material.fragmentShader
+				}
 
-		}
+			}
 
-		var program;
+			// Check if code has been already compiled
 
-		// Check if code has been already compiled
+			for ( var p = 0, pl = _programs.length; p < pl; p++ ) {
 
-		for ( var p = 0, pl = _programs.length; p < pl; p ++ ) {
+				var programInfo = _programs[p];
 
-			var programInfo = _programs[ p ];
+				if ( programInfo.code === code ) {
 
-			if ( programInfo.code === code ) {
+					program = programInfo;
+					program.usedTimes++;
 
-				program = programInfo;
-				program.usedTimes ++;
+					break;
 
-				break;
+				}
 
 			}
 
-		}
+			material.program = program;
+
+			if ( program === undefined ) {
 
-		if ( program === undefined ) {
+				program = new THREE.WebGLProgram( _this, code, material, parameters, _shaderCache );
+				_programs.push( program );
 
-			program = new THREE.WebGLProgram( _this, code, material, parameters );
-			_programs.push( program );
+				_this.info.memory.programs = _programs.length;
 
-			_this.info.memory.programs = _programs.length;
+				material.program = program;
 
+				return false;
+			}
 		}
 
-		material.program = program;
+		program = material.program;
+
+		if ( program.state !== THREE.WebGLProgram.LinkedState ) {
```

Comment:
Would need to be ```THREE.WebGLProgram.CompileErrorState || THREE.WebGLProgram.LinkErrorState``` as it could also be in ```THREE.WebGLProgram.LinkingState``` or ```THREE.WebGLProgram.CompilingState``` but could probably put something like that in.

Also would need to check size of position attribute to see if its 3 or 4 for the error shader.